[telemetry] Detect service reachability issues

muhamadazmy opened the pull request:
[telemetry] Detect service reachability issues

Summary:
Introducing a counter for number of "rechability" issues for a service
that can detect a service is down or un-responsive

by visualizing `rate(restate.invoker.deployment_unreachable_errors.total)` coupled with alerts, operator
can know when a service is facing connectivity problems

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/3426).
* __->__ #3426
* #3439

pcholakov commented on crates/invoker-impl/src/metric_definitions.rs:
```diff
@@ -18,7 +18,10 @@ pub const INVOKER_INVOCATION_TASKS: &str = "restate.invoker.invocation_tasks.tot
+pub const INVOKER_SERVICE_RESPONSE_TIME: &str = "restate.invoker.service_response_time.seconds";
 pub const INVOKER_TASKS_IN_FLIGHT: &str = "restate.invoker.inflight_tasks";
+pub const INVOKER_JOURNAL_REPLAY_TIME: &str = "restate.invoker.journal_replay_time.seconds";
+pub const INVOKER_SERVICE_DOWN_ERRORS: &str = "restate.invoker.service_down_errors.total";
```
Nitpicky naming observation: Wondering if "unavailable" or "unreachable" might not be more accurate than "down" - since we can't tell authoritatively that it's really down, just that it's not available from our point of view.

muhamadazmy replied:
Ah very nice. Thank you. Will apply :)

pcholakov reviewed (approved):
This is a great observability addition, thank you @muhamadazmy! Left one minor naming comment which you should feel free to ignore.

AhmedSoliman commented on crates/invoker-impl/src/metric_definitions.rs:
```diff
@@ -21,6 +21,8 @@ pub const INVOKER_TASK_DURATION: &str = "restate.invoker.task_duration.seconds";
 pub const INVOKER_TASKS_IN_FLIGHT: &str = "restate.invoker.inflight_tasks";
 pub const INVOKER_JOURNAL_REPLAY_TIME: &str = "restate.invoker.journal_replay_time.seconds";
+pub const INVOKER_SERVICE_UNREACHABLE_ERRORS: &str =
+    "restate.invoker.service_unreachable_errors.total";
```
I guess you mean deployment

AhmedSoliman commented on crates/invoker-impl/src/lib.rs:
```diff
@@ -1090,6 +1090,11 @@ where
             debug_assert_eq!(invocation_epoch, ism.invocation_epoch);
+
+            if self.is_service_down_error(&error) {
+                counter!(INVOKER_SERVICE_UNREACHABLE_ERRORS, "service" => ism.invocation_target.service_name().to_string()).increment(1);
```
I think we need to have both the deployment id and the service name. The risk is that it this will be a high cardinality metric.

muhamadazmy committed:
[invoker] Attach deployment id to the invoker error

This is used in the next PR to associate the deployment unreachable
failure with the deployment id

muhamadazmy committed:
[telemetry] Detect service reachability issues

Summary:
Introducing a counter for number of "rechability" issues for a service
that can detect a service is down or un-responsive

by visualizing `rate(restate.invoker.deployment_unreachable_errors.total)` coupled with alerts, operator
can know when a service is facing connectivity problems

pcholakov reviewed (approved):
Nice one, thanks @muhamadazmy!

AhmedSoliman commented on crates/invoker-impl/src/lib.rs:
```diff
@@ -1090,6 +1093,15 @@ where
+                    .deployment_id
+                    .map(|id| Cow::Owned(id.to_string()))
+                    .unwrap_or_else(|| Cow::Borrowed("unknown"));
+                counter!(INVOKER_DEPLOYMENT_UNREACHABLE_ERRORS, "service" => ism.invocation_target.service_name().to_string(), "deployment" => deployment_id).increment(1);
```
Is this something we can already detect with the new fttb metric, or the invocation task status metric or does it require a new metric?
